[WIP] Traktor Kontrol S4mk3: motors revisited

jtMUMT opened the pull request:
Revising the motorized platter control on the Traktor S4Mk3 controller.

Current Status: motors are usable for mixing and work almost identically to using Traktor. Code needs major cleanup.

Impact on common code: removed the 20ms cap on QTimer requests so that the motors can be controlled at the correct sampling rate to match the sensor input stream (2ms / 500Hz).

~~Extraneous / "bonus" changes: added midpoint offset for tempo faders. Not implemented in the most graceful way but it works.~~ Implemented upstream already

Known bugs: switching decks on the right side seems to change the pad mode / screw up the LED colours on that side. I suspect it has to do with how I'm assigning left/right status to each deck.

Apologies: for all the hacking and rearranging/renaming of existing declarations. This is very much a work in progress and the code needs to be revised before it's ready to merge (assuming it passes review of course). I have ambitions to improve the overall comprehensiveness of the mapping file by fixing hardcoded values and harmonizing some of the nomenclature. So far this has resulted in some improvements and also some unintended ambiguities (eg., I didn't realize that "slip" mode was already a thing before I used this to refer to the slipmat emulation). Right now this code is here so that others can try it out while I/we keep working on it :)

Commit:
branching this out to keep my comments etc as I learn about the motor control code

Commit:
Merge remote-tracking branch 'origin/main' into s4m3_motors_revisited

Commit:
saving changes due to controller errors

Commit:
merged with more recent changes to main

Commit:
Implemented input filter buffer (untested)

Commit:
Implemented new output motor torque with PI controller and smoothing filter

Commit:
tuning the PID controller which is starting to work very well!

Commit:
playback quantizing experiments

Commit:
playback quantization works now a bit like traktor, but need to analyze the traktor response to nudging (too easy to stop the platter)

Commit:
motor testing for determining output->input mapping

Commit:
further motor output-input testing

Commit:
trying to fix the nudge mode but i'm not doing very well

Commit:
things are not working

Commit:
ok it's getting there, but right now the "rate" setting that I'm using is competing with the tempoFader so it's not ideal. but otherwise the quantized playback is working really well!

Commit:
workable! jog setting is nice. still some work to do so that you can eg., switch decks without affecting the timing

Commit:
added zero point calibration to tempo faders (to pot class generally)

Commit:
fixed switch deck functionality with turntable mode on

Commit:
tempofader centering works as intended now

Commit:
fixed a hardcoded value

jtMUMT commented:
Correction: there is no bug with the pad LEDs. That was simply human error (I set the default pad mode by mistake)

JoergAtGithub commented:
Welcome at Mixxx!
As a first-time contributor we need you to sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

acolombier commented:
Hi @jtMUMT , thanks for looking into this. It looks like you based your work off an old version of the mapping. Could you please move you changes off the latest version in `main`?

ronso0 commented:
Also, please install `pre-commit`
https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking
and run it on the changed files. After rebase onto 2.5, I _think_ this can be done with
```
FILES=$(git log --name-only --pretty="" mixxx/2.5..HEAD)
pre-commit run --files $FILES
```
or, if you have `xargs`
`git log --name-only --pretty="" mixxx/2.5..HEAD | xargs pre-commit run --files`

acolombier commented:
Note that another way to run pre-commit after pushing is the following:

```sh
pre-commit run --show-diff-on-failure --color=always --from-ref upstream/2.5 --to-ref HEAD
```

ronso0 commented:
@jtMUMT Could you share the settings for your preferred mix/scratch experience?

jtMUMT commented:
@acolombier thanks for the quick review. I'm going to redo my changes from the base file again, only changing what's necessary and keeping the existing naming conventions and style as you indicate. The only variables I would still want to rename are the ones that I consider strongly confusing, specifically when calculating the position derivative.

@ronso0 please see my previous comment for recommended settings. I'll copy them again here in case my previous post doesn't show for some reason. I can explain in detail any of the effects of these values and in the final version it will be as comprehensive as possible.

> here are the controller values (set in preferences) that work for me. I'll make these default in a future commit :)
> 
> Jogwheel Tension 0.5
> Jogwheel Torque 60000
> P gain 80000
> I gain 500
> D gain 50000
> Slip Force 12000
> Motor Friction Compensation 0

I hope your motors are OK too! The last thing I want is to cause hardware problems. My guess is that the sound you were hearing was due to low output values; insufficient current to get the motor turning.

jtMUMT commented:
Actually, since you mention second-hand wheels, a note: the 'nudge' functionality is currently calibrated to 33.3RPM motor output measured on my hardware, so it may run at a slightly higher or lower pitch than expected if your motors have a significantly difference in current to rpm. Please don't use 45rpm mode yet, as the mapping is currently hardcoded for 33.3 for the nudge mode. Nothing will break, it will just misinterpret the velocity data for nudging calculations.

I'm not sure if this will become a sticking point, or how much variation there is between motors. It is possible to automatically self-calibrate so the user doesn't have to worry about this stuff, but it's something to consider a few steps down the line, once everything is working with manual calibration.

Commit:
Added motor setting defaults to preferences

Commit:
Fix: added output target for 45rpm, removed friction compensation setting

Commit:
(untested) style: undoing some of the rearrangements I made to const declaration blocks etc, as requested

Commit:
(untested) style: renaming consts and vars to style guide, cleaning comments

Commit:
(untested) removed tempo fader offset code

Commit:
Continuing to clean up the code and rename vars

jtMUMT commented on res/controllers/Traktor-Kontrol-S4-MK3.js:
```diff
@@ -1510,7 +1832,37 @@ class S4Mk3EffectUnit extends ComponentContainer {
+// This is a global variable so the S4Mk3Deck Components have access
+// to it and it is guaranteed to be calculated before processing
+// input for the Components.
+let wheelTimerDelta = 0; if (this.shifted && this.unit.focusedEffect === null) {
```
Agreed. This was actually me putting those lines back to where they were before my branch, responding to @acolombier
> * It would be great if you could leave the change to a minimal and restore changes unrelated in this context (e.g moving colour constant around

I'd be happy to better-organize the definitions but I don't want to overstep the scope of this pull request. Would it be better to do a big cleanup after the new feature/fix is integrated and in good working order?

ronso0 commented:
So the rebase went well?
(I'm hesitating to switch to 2.6 here, last time that did mess things up, idk)

To reduce conflicts (now and later on) I suggest to remove changes unrelated to motors, like the tempo fader stuff which is already part of #14882

jtMUMT commented:
Yup, I've removed the tempo fader code already. I had to abort the rebase
unfortunately. It was looking like it would cause havoc and i got scared

On Fri, Jun 20, 2025, 19:25 ronso0 ***@***.***> wrote:

> *ronso0* left a comment (mixxxdj/mixxx#14872)
> <https://github.com/mixxxdj/mixxx/pull/14872#issuecomment-2991092698>
>
> So the rebase went well?
> (I'm hesitating to switch to 2.6 here, last time that did mess things up,
> idk)
>
> To reduce conflicts (now and later on) I suggest to remove changes
> unrelated to motors, like the tempo fader stuff which is already part of
> #14882 <https://github.com/mixxxdj/mixxx/pull/14882>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/14872#issuecomment-2991092698>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABU42XEKO4QVGOM4RISPWFT3EPVTTAVCNFSM6AAAAAB6HARDO6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSOJRGA4TENRZHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jtMUMT commented:
> Okay, cool! I gave this another spin and it feels nice. And the motors are quiet when playing!
Thank you for testing! I'm glad it's working and extra glad your motors are OK after last time 😅 

> * a few times I managed to put it in to a state where `scratch2` was not reset on wheel release, so even after I hit pause, the track kept "playing" and wheel kept spinning
Noted, I'll look into it as soon as I can. To be clear, the sequence was: 1) engage scratch mode sometime during playback; 2) stop scratching but continue playback; 3) hit pause, but playback continues?

> * when scratching back and forth I feel this stutter effect like when going over the Cue. I didn't scratch particularly fast though, just a slow baby scratch
Thinking about it, I suspect I know the cause of this---when the velocity gets close to the target velocity the motor controller might be trying to assert more control to snap it onto target. Should be easily fixed. 

> * when stopping the wheel with the crown, the deck doesn't stop. I understand this shouldn't behave like when touching the top, but when stopped with the crown it should stop after a while
This might be largely a preference thing but definitely worth making it tweakable. The way I've tuned it right now was to make it usable for a gig I had coming up, but I agree it can use some adjustment. One question mark though, on a real turntable do we ever stop the platter by dragging on the crown?

ronso0 commented:
> .. and extra glad your motors are OK after last time 😅

Hahaa, me too :sweat_smile: 

> > * a few times I managed to put it in to a state where `scratch2` was not reset on wheel release, so even after I hit pause, the track kept "playing" and wheel kept spinning
> 
> Noted, I'll look into it as soon as I can. To be clear, the sequence was: 1) engage scratch mode sometime during playback; 2) stop scratching but continue playback; 3) hit pause, but playback continues?

I think it was after I tried to stop the deck with the crown, not sure though.

________
> > * when stopping the wheel with the crown, the deck doesn't stop. I understand this shouldn't behave like when touching the top, but when stopped with the crown it should stop after a while
> 
> This might be largely a preference thing but definitely worth making it tweakable. The way I've tuned it right now was to make it usable for a gig I had coming up, but I agree it can use some adjustment. One question mark though, on a real turntable do we ever stop the platter by dragging on the crown?

Thanks.
Idk, I sometimes do weird stuff with my turntable (Reloop RP-7000 Mk2, it has +/-50% ultra-pitch and 78 RPM).
And since the crown and the platter are one part physically on the S4, too, I'd just expected it to stop the deck in TT mode :shrug:

ronso0 commented:
I'll try to put this into my custom mapping (using `scratch_position` instead of `scratchTick`), but for proper testing I'll wait for fixup for 
> when scratching back and forth I feel this stutter effect like when going over the Cue.

JoergAtGithub commented on src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp:
```diff
@@ -509,10 +509,10 @@ int ControllerScriptInterfaceLegacy::beginTimer(
-    if (intervalMillis < 20) {
+    if (intervalMillis < 2) {
         qCWarning(m_logger) << "Timer request for" << intervalMillis
-                            << "ms is too short. Setting to the minimum of 20ms.";
-        intervalMillis = 20;
+                            << "ms is too short. Setting to the (hacked) minimum of 2ms.";
+        intervalMillis = 2;
```
This can't be guaranteed on al host operating systems now. And controller mappings must work on all operating systems.

ronso0 replied:
FYI Qt 6.8 introduces QChronoTimer https://doc.qt.io/qt-6/qchronotimer.html
> For [Qt::PreciseTimer](https://doc.qt.io/qt-6/qt.html#TimerType-enum), QChronoTimer will try to keep the precision at 1ns.

JoergAtGithub replied:
I fear this much more complex, you must tell the operating systems, that the system wide event scheduler must not throttle down for energy savings.
This is not possible with Qt commands to my knowledge. This requires native OS commands like https://learn.microsoft.com/en-us/windows/win32/api/timeapi/nf-timeapi-timebeginperiod on Windows (which I tried btw., and didn't work for me).
On my system (I guess it's the same on all Windows 11 systems) the minimum period is 15.5ms for example.

JoergAtGithub replied:
HID input reports are buffered in a FIFO running inside the operating systems kernel. Normally, we process them rather fast and there are seldom more than two reports in this buffer. It's safe to assume, that reports are never skipped, due to buffer overload.
If we assume, that there is a realtime timer in the controller, we can use the average time between two reports to determine the time, when the report was sent.
Please note, that sending HID output reports can take much longer than receiving input reports.

Commit:
corrected location of wheel mode values. minor edits to comments

jtMUMT commented:
RE: playback continues, but wheel stops spinning, when play button is pressed to stop playback
@ronso0 I have confirmed that this was not an issue in commit `de60ce31cd8d9685e8f1e06e74aac40ae292f547`  so I must have introduced further bugs in my recent attempt to clean up the code.

Commit:
fixed issue with playback continuing after pausing

Commit:
adjusted default motor control parameters

jtMUMT commented:
@ronso0 the issue with playback continuing should be fixed now. This may have also fixed the 'bump' you were describing during scratching. Please let me know if/when you have a chance to try it out. In the meantime I'll turn my attention to the HID output as discussed above.

Commit:
calling motor outputs from the HID input method to avoid QTimer limitations

jtMUMT commented:
@JoergAtGithub I moved the motor output call to the HID input handler, and perceived no subjective difference in performance. Seems like a winner, assuming it works as well on different platforms

Commit:
restore QTimer limit to 20ms

JoergAtGithub commented:
Please revert the remaining change in controllerscriptinterfacelegacy.cpp too.
There is now a merge conflict, as another PR against the S4 Mk3 mapping got merged. This needs to be resolved manually, berfore this PR can be merged.

Commit:
minor changes to comments

Commit:
Moved nudge sensitivity to prefs; cleaned up comments and const names

Commit:
Further cleanup of code comments

Commit:
finished cleaning up comments

jtMUMT commented:
I can't see any other changes to controllerscriptinterfacelegacy.cpp in my local repo---it appears to me like it's in the same state it was before I started my branch. 

I'll manually merge the new mapping file at my next opportunity!

JoergAtGithub commented:
Try to merge the target of this PR into your local branch. Than you will have to resolve the conflicts and afterwards you should see the difference. Commands should be like these, depending how you named upstream:
```
git fetch upstream
git merge upstream/main
```

Commit:
minor pre-merge modifications

Commit:
Merge commit 'aae018e1bba90a39670bd02641d6e8f74d3f1d62' into s4m3_motors_revisited

ronso0 commented:
I merged this into my mapping, and while scratching seems to work as before :tada:  I again ran into the 'deck keeps playling / scratch2 not reset'.
I _think_ it happened when playing with the settings (or after editing the script?) while the deck was playing so it might be related to init/shutdown cleanup?

Will test the pure PR version later today.

jtMUMT commented:
Thanks again for testing :)

Scratching near the tempo it's exactly as you say, the motor is switching scratch mode on and off. With a real slipmat it should have the least resistance when near the motor speed! I'll modify the behaviour to add a dead zone to the slipmat resistance, and to only disable scratch mode once the performer stops touching the disc.

Thanks for reminding me of the lack of reverse. Believe it or not Traktor doesn't spin in reverse even with the REV button on. I can guess based on other compromises they've made to the "pure" turntable emulation (such as the crown nudging and instant spin-up), they want the REV playback function to kick in instantly without having to spin down/up the platter. And likewise so that when REV is released, playback resumes without delay. 

So again, like the question of whether manipulating the crown should directly map to the playback speed instead of temporarily nudging it, this is a design choice about what is the priority. Should TT mode compromise towards instant REV response so that it acts more like a CDJ, or should it be as faithful as possible to real TT behaviour? 

If the former: spinning the platter in reverse is largely aesthetic and introduces complexity without necessarily improving functionality. 

If the latter: fair play, I like turntables too! But how does this clash with FLUX mode, for example?

Maybe a future ideal would be to have an option to enable "real" TT emulation vs "compromised" emulation.

On some level, we do have to admit this is not a turntable and has different physical dynamics. But as a turntable lover I'd be so happy to have something that feels just like the real thing :)

jtMUMT commented:
Inpropose that for this PR we focus on matching Trakor's implementation, and then start a new feature afterwards to bring in our own TT simulation adjustments. This way it's as accessible as possible to existing users of the controller who are accustomed to Traktor.

Commit:
Post-merge cleanup

fixed a mistake during manual merge, addressed many pre-commit warnings and errors

Corrected error in the TempoFader calibration settings: missing unit conversion from mm to ticks. Added proper warning message for incorrect deck assignment

Requested: reverted Tempo Fader changes

typo: delete line of tabs

reverted comment in scriptinterfacelegacy.cpp

removed unused var oldWheelTimer

Commit:
Jogwheel TT mode fixes and improvements

added dead zone during scratching to avoid 'chattering' motor problem

re-enabled scratch mode on TT mode activation: fixes issue where scrubbing was not possible using crown adjustment until disc was touched

fixed error when selecting TT mode; scratch mode will not be re-enabled if playback is already underway

Fixed: inability to scrub with crown while paused in motor mode (faulty conditional)

Fixed: selects proper rotation speed (more faulty conditionals). Tweak to default proportional error gain

jtMUMT commented:
@ronso0 deadzone for scratching should now reduce motor chatter considerably. Hand-spinning feels much better to me

@JoergAtGithub I've corrected an error I made during the merging of the new TempoFader feature, and have also corrected an error in the implementation of that feature (was missing a unit conversion). Everything should be in good working order now, pending any testing issues or modification requests from @ronso0

JoergAtGithub commented:
There is still a changed comment in the .cpp file left, please restore the original state of this file:
https://github.com/mixxxdj/mixxx/pull/14872/files#diff-89123d0edabf461103cd3837c2b1e4c362da8c533781aa9e0ff1032de4cd2e9d

ronso0 commented:
Please revert the tempo fader changes. Related changes have been merged to 2.5 recently and will be in main soon, too.

ronso0 commented:
> trigger engine.brake() and engine.softStart() (reverse in the REV case) with parameters (roughly) matching the motor timing.

Oh, not possible atm. `softSart()` only reads the current rate (tempo slider) and can only play forward.
(though I'm testing if this can consider `reverse`)

acolombier committed:
S4MK3: refactor device constant

acolombier committed:
S4MK3: improve stability with CO are missings

acolombier committed:
S4MK3: use button hooks and add shared data API

acolombier committed:
S4MK3: add stem and improve hotcues

acolombier committed:
S4MK3: add beat jump pad

Commit:
Feature: reverse motors when REV button is held

acolombier committed:
chore: add missing global in ESlint

ronso0 commented:
Thank you for the fixups :+1: 
I'll try this again at the end of the week (I hope..)

Unfortunately coflicts have emerged, probably because the tempo fader center fix has landed in `main`.
Btw I still think --as this is a mapping improvement that doesn't require new `main` features-- this should target 2.6 (actually 2.5) so the new mapping is available asap.
Can you try to rebase onto 2.6 -- and while you do so, try to squash commits (especially the tempp fader do/undo)?

While you're at it (I guess this is quite some work, maybe not) I'm still in favor of _not_ reordering constants here. "Just" the motor mapping.
Not sure if you're aware, there's also a mapping extension #15216 (excerpt from original #13653) and a screen mapping #13653 in progress and it would save you / @acolombier the hazzle of resolving conflicts.

jtMUMT commented:
Thanks for the replies! I'm aware of the coming screen support, which I'm very excited about. Didn't know about the mapping extension though---the S4 is gonna be in really great shape soon :smiley: 

Here's what I'll do:
* Squash minor commits to clean up branch commit history
* Remove and/or simplify motor control preferences (there's really no need for regular users to mess with the PID gains, but the nudge sensitivity and slipmat friction can stay)
* Rebase onto 2.6

@ronso0 I already reverted unnecessary reordering of constants a while back, so no worries there

Commit:
Usability: remove PID controller coefficients from user preferences

Commit:
Merge branch 'feat/s4-mk4-improvements' into s4m3_motors_revisited

Commit:
Reverting and correcting differences in const and var names: trying to make my new code more in line with the existing base even if there are some historical deviations from the style guide that need fixing (those can be fixed in a future PR)

jtMUMT commented:
- I've squashed the recent commits to my branch so that they're more concise and clean.
- Motor controller preferences are simplified. 
- Merged @acolombier 's S4Mk3 improvements branch into my PR to bring it up to date with 2.6*

*I tried rebasing but it would have necessitated 62 commits' worth of manual merges and I felt this would be too likely to introduce further errors.

ronso0 commented:
> the motor chattering is back. sometimes right away when I enable TT mode, but almost always after scratch release

I think I fixed that.

The cue/hotcue goto_and_stop/preview issue is a bit annoying.
I disabled that, but this also disables  Vinyl stop which is actualy nice : )

Btw I think there should be some way to reset() the motor manager when switching to/from motor mode.
This would allow reset in case  of issues like the chattering after slipmat mode.

jtMUMT commented:
Yeah when I try to make the changes, the pre-commit is blocking me:

```
 363:30  error  Expected '===' and instead saw '=='  eqeqeq

3796:57  error  Expected '===' and instead saw '=='  eqeqeq

3853:42  error  Expected '===' and instead saw '=='  eqeqeq
```

I'm sure this is down to my own unfamiliarity with JS and these conditionals could be improved, but I'd love to know what exactly is happening. Seems like instance variable always return `true` even if they're booleans set to `false`
